File: backend/core/losses.py
import numpy as np

### Tính loss phục vụ cho việc hiển thị ###

_EPS = 1e-9

def mse_loss(y_pred, y_true):
    return np.mean((y_pred - y_true) ** 2)

def bce_loss(y_pred, y_true):
    return -np.mean(y_true * np.log(y_pred + _EPS) + (1 - y_true) * np.log(1 - y_pred + _EPS))

def ce_loss(y_pred, y_true):
    return -np.mean(np.sum(y_true * np.log(y_pred + _EPS), axis=-1))

### Tính đạo hàm cho layer cuối ###

def bce_ouput_grad(y_pred, y_true):
    n = y_pred.shape[0]
    return (y_pred - y_true) / n

def ce_output_grad(y_pred, y_true):
    n = y_pred.shape[0]
    return (y_pred - y_true) / n

def mse_output_grad(y_pred, y_true):
    n = y_pred.shape[0]
    return 2 * (y_pred - y_true) / n

File: netlab/src/engine/losses.ts
import { meanAll, zipMatrix } from './matrix'

const EPS = 1e-9

export function mseLoss(yPred: number[][], yTrue: number[][]): number {
    return meanAll(zipMatrix(yPred, yTrue, (p, t) => (p - t) ** 2))
}

export function bceLoss(yPred: number[][], yTrue: number[][]): number {
    const terms = zipMatrix(yPred, yTrue, (p, t) => t * Math.log(p + EPS) + (1 - t) * Math.log(1 - p + EPS))
    return -meanAll(terms)
}

export function ceLoss(yPred: number[][], yTrue: number[][]): number {
    const perRow = zipMatrix(yPred, yTrue, (p, t) => t * Math.log(p + EPS)).map((row) => row.reduce((a, b) => a + b, 0))
    return -(perRow.reduce((a, b) => a + b, 0) / perRow.length)
}

export function bceOutputGrad(yPred: number[][], yTrue: number[][]): number[][] {
    const n = yPred.length
    return zipMatrix(yPred, yTrue, (p, t) => (p - t) / n)
}

export function ceOutputGrad(yPred: number[][], yTrue: number[][]): number[][] {
    const n = yPred.length
    return zipMatrix(yPred, yTrue, (p, t) => (p - t) / n)
}

export function mseOutputGrad(yPred: number[][], yTrue: number[][]): number[][] {
    const n = yPred.length
    return zipMatrix(yPred, yTrue, (p, t) => (2 * (p - t)) / n)
}
